File: HelloWorldiOS/ViewController.m
//
//  ViewController.m
//  HelloWorldiOS
//
//  Created by Anton Chuev on 2/12/18.
//  Copyright © 2018 Applitools. All rights reserved.
//

#import "ViewController.h"
#import "BEMCheckBox.h"
#import "InitialStrategy.h"
#import "DifferencesStrategy.h"

static int GenerateRandomNumberCheckboxTag = 0xAAAA;
static int SimulateDifferencesCheckboxTag = 0xBBBB;

@interface ViewController () <BEMCheckBoxDelegate>

@property (weak, nonatomic) IBOutlet UILabel *helloLable;

@property (weak, nonatomic) IBOutlet BEMCheckBox *generateRandomNumberCheckbox;
@property (weak, nonatomic) IBOutlet BEMCheckBox *simulateDifferencesCheckbox;

@property (weak, nonatomic) IBOutlet UILabel *randomNumberLabel;
@property (weak, nonatomic) IBOutlet UIButton *clickMeButton;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *clickMeButtonCenterXLayout;
@property (weak, nonatomic) IBOutlet UILabel *buttonClickedLable;
@property (weak, nonatomic) IBOutlet UIImageView *imageView;

@property (assign, nonatomic) BOOL clickMeContentShown;
@property (strong, nonatomic) id <Strategy> currentStrategy;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.generateRandomNumberCheckbox.tag = GenerateRandomNumberCheckboxTag;
    self.simulateDifferencesCheckbox.tag = SimulateDifferencesCheckboxTag;
    
    [self updateRandomNumberLable];
    
    self.currentStrategy = [InitialStrategy new];
    self.helloLable.attributedText = [self.currentStrategy attributedText];
    
    self.clickMeContentShown = NO;
    [self showClickMeContent:self.clickMeContentShown];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    self.clickMeButton.layer.cornerRadius = 5.f;
    [self updateContentWithStrategy:self.currentStrategy];
}

- (void)showClickMeContent:(BOOL)shouldShow {
    __weak typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        weakSelf.buttonClickedLable.hidden = !shouldShow;
        weakSelf.imageView.hidden = !shouldShow;
    });
}

- (void)updateContentWithStrategy:(id<Strategy>)strategy {
    __weak typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        weakSelf.imageView.image = [strategy image];
        weakSelf.clickMeButtonCenterXLayout.constant = [strategy clickMeButtonCenterXLayoutValue];
        weakSelf.helloLable.attributedText = [strategy attributedText];
    });
}

- (NSInteger)generateRandomInteger {
    return rand()%NSIntegerMax+1;
}

- (void)updateRandomNumberLable {
    NSInteger randomInteger = [self generateRandomInteger];
    __weak typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        weakSelf.randomNumberLabel.text = [NSString stringWithFormat:@"%li", (long)randomInteger];
    });
}

#pragma mark

- (IBAction)clickMeButtonTapped:(id)sender {
    self.clickMeContentShown = !self.clickMeContentShown;
    [self showClickMeContent:self.clickMeContentShown];
}

#pragma mark - BEMCheckBoxDelegate

- (void)didTapCheckBox:(BEMCheckBox *)checkBox {
    if ( checkBox.tag == GenerateRandomNumberCheckboxTag ) {
        if ( checkBox.on ) {
            [self updateRandomNumberLable];
        }
    } else {
        if ( checkBox.on ) {
            self.currentStrategy = [self.currentStrategy isKindOfClass:[InitialStrategy class]] ? [DifferencesStrategy new] : [InitialStrategy new];
            [self updateContentWithStrategy:self.currentStrategy];
        } else {
            self.clickMeContentShown = NO;
            [self showClickMeContent:self.clickMeContentShown];
        }
    }
}

@end

File: HelloWorldiOS/ViewController.swift
// Licensed under the Applitools SDK License, which can be found here: https://www.applitools.com/eula/sdk

import UIKit

class ViewController: UIViewController {

    @IBOutlet weak var helloLabel: UILabel!
    @IBOutlet weak var randomCheckbox: UIButton!
    @IBOutlet weak var randomNumberLabel: UILabel!
    @IBOutlet weak var tapMeButtonCenterXLayout: NSLayoutConstraint!
    @IBOutlet weak var bottomContainer: UIView!
    @IBOutlet weak var imageView: UIImageView!

    @IBOutlet weak var simulateButton: UIButton!
    
    private var model = Model()
    private var tapMeContentShown = false {
        didSet {
            showTapMeContentIfNeeded()
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        updateContent()
    }

    @IBAction func didTapGenerateRandomNumber(checkbox: UIButton? = nil) {
        randomCheckbox.isSelected.toggle()
        model.isNumberRandom.toggle()
        updateNumberLabel()
    }

    @IBAction func didTapSimulateDifferences(checkbox: UIButton? = nil) {
        simulateButton.isSelected.toggle()
        model.hasDifferences.toggle()
        tapMeContentShown = false
        updateContent()
    }

    @IBAction func tapMeButtonTapped(sender: UIButton? = nil) {
        tapMeContentShown.toggle()
    }

}

private extension ViewController {

    func updateNumberLabel() {
        randomNumberLabel.text = String(model.number)
    }

    func updateContent() {
        helloLabel.attributedText = model.attributedText
        updateNumberLabel()
        tapMeButtonCenterXLayout.constant = model.tapMeButtonCenterXLayoutValue
        imageView.image = UIImage(named: model.imageName)
        showTapMeContentIfNeeded()
    }

    func showTapMeContentIfNeeded() {
        bottomContainer.isHidden = !tapMeContentShown
    }

}
